inst/tutorials/practical0-1/practical0-1.Rmd

```diff
@@ -0,0 +1,444 @@
+happens outside the function. Unfortunately, the reverse isn't true. Look at
+this function:
+
+```{r global, exercise.cap = "Exercise 3", exercise=TRUE}
+missing_add <- function(first) {
+  first + second
+}
+
+missing_add(10)
+```
+
+This function was intended to be like `add_up()` in *Exercise 2*, but I
+accidentally forgot to add the second argument. Fortunately if you run it, it
+will give an error (try it). However, what happens if you have defined
+a variable called `second` elsewhere in your script? Try adding the following
+at line 4:
+
+```{r, eval=FALSE}
+second <- 1
+```
+
+`missing_add(10)` should no longer give an error, and return `11`. Try changing
+`second` to 2 - `missing_add(10)` should now return `12`. That's not good news -
+what the function does depends on what else you may have done in your script.
+When you call a function it should always do the same if you give it the same
+arguments.
+
+This gets even worse when you account for our ability to misspell things:
+
+```{r misspelling, exercise.cap = "Exercise 3", exercise=TRUE}
+misspelled_add <- function(first, second) {
+  resuIt <- first + second
+  result
+}
+
+result <- -5
+misspelled_add(10, 20)
+misspelled_add(1, 2)
+```
+
+<div id="misspelling-hint">
+**Hint:** If you can't see the problem see how the word `result` is spelled in
+different places.
+</div>
+
+You should find that `misspelled_add()` gives the wrong answer even though the
+code looks right (if you don't look too closely). This problem can (and does!)
+cause hours of agony to some students every year on this course. As a result,
+we insist on two things in your submitted practicals:
+
+1. You use variable names inside functions that you don't use outside in your
+   scripts wherever possible. Sometimes there really only is one sensible name
+   for a variable (maybe `timestep` when describing the increments in your
+   simulation model), and you should be very careful when that happens, but
+   normally there are at least two ways of describing the same idea - often one
+   more general in a function (`birth.rate`) which can be used in multiple
+   contexts, and then specific in your main script each time you use it
+   (`human.annual.birth.rate`, for instance).
+2. You write checks for your functions to make sure they aren't accidentally
+   missing any variables or arguments. There are two tools for this,
+   `findGlobals()` and `checkUsage()` in the `codetools` library.
+
+Try adding the following lines to the end of *Exercise 3*:
+
+```{r, eval=FALSE}
+library(codetools)
+checkUsage(misspelled_add)
+```
+
+You'll see that it reports on the problem that you have probably already
+identified. It can also identify other problems, and we recommend it to you.
+However, it doesn't return anything you can use in your code to allow us to
+automatically reject bad functions, so there is a second solution we'll
+use in our code, and we'll show you how to use it in a later exercise, but
```

Comment:
automatically reject bad functions, ~~so there is a second solution we'll
  use in our code, and~~ we'll show you how to ~~use it~~ [do this] in a later exercise, ~~but~~